doc: add a guide to host a static website using Hugo + cellar (#349)
CleverCloud/documentation: content/guides/hugo-static-s3.md

---
title: Static hosting with Cellar
description: Build your website with Hugo, then deploy it through a S3-compatible object storage, Cellar. 
tags:
- guides
keywords:

type: "docs"
comments: false
draft: false
---

> When creating a static website, it's possible to host it using a runtime.

Reviewer: ```suggestion

When creating a static website, it's possible to host it using a runtime.
```

> You can also also use Cellar, a S3-compatible object storage. This hosting method saves time on configuration and application management, and is fast.

Reviewer: ```suggestion
You can also use Cellar, a S3-compatible object storage. This hosting method saves time on configuration and application management, and is fast.
```

> To do so, you will need:

Reviewer: ```suggestion
To do so, you need:
```

- a [Clever Cloud account](https://developers.clever-cloud.com/doc/quickstart/)
- [Hugo](https://gohugo.io/) on your machine
- to have [created your website](https://gohugo.io/getting-started/quick-start/) files with Hugo
- a Cellar add-on; if you have never created an add-on, you can follow [this guide](https://developers.clever-cloud.com/doc/quickstart/#create-your-first-add-on)

## Bucket creation and management

To create a bucket:

> - you can follow the [bucket creation page](https://developers.clever-cloud.com/doc/addons/cellar/#creating-a-bucket),

Reviewer: ```suggestion
- Follow the [bucket creation instructions](https://developers.clever-cloud.com/doc/addons/cellar/#creating-a-bucket),
```

Now, you can set the policy to your bucket using s3cmd:

```bash
s3cmd setpolicy ./policy.json s3://<bucket-name>
```

If you encounter errors, you might need to specify the [configuration file path](#download-the-configuration-file):

```bash
s3cmd setpolicy ./policy.json -c path/to/s3cfg.txt s3://<bucket-name>
```

All of your objects should now be publicly accessible.
If needed, you can delete this policy by using:

```bash
s3cmd delpolicy s3://<bucket-name>
```

The original ACL should apply to all of your objects after that.

## Hugo configuration for Cellar

Hugo doesn’t know by default the path to your bucket in your Cellar. You have to add a few things to guide the deployment process.

In the folder of your website, you’ll find Hugo’s configuration file:

- `hugo.toml`
- `hugo.yaml`
- `hugo.json`

Open it, then add the following (according to the programing language):

{{< tabs items="JSON,YAML,TOML" >}}

  {{< tab >}}**JSON**

```json {filename="hugo.json"}
  {
    "deployment": {
      "targets": [
        {
          "name": "(a name for your own reference)",
          "URL": "s3://<BUCKET_NAME>?endpoint=https://<CELLAR_HOST>&region=fr-par"
        }
      ]
    }

>   }

Reviewer: ```suggestion
  }
    ```
 
```

{{< /tab >}}

  {{< tab >}}**YAML**

>   ```json {filename="hugo.yaml"}

Reviewer: ```suggestion

  ```json {filename="hugo.yaml"}
```

  deployment:
  targets:
    - name: "(a name for your own reference)"

>     - URL: "s3://<BUCKET_NAME>?endpoint=https://<CELLAR_HOST>&region=fr-par"

Reviewer: ```suggestion
    - URL: "s3://<BUCKET_NAME>?endpoint=https://<CELLAR_HOST>&region=fr-par"
    ```

```

  {{< /tab >}}
  
  
  {{< tab >}}**TOML**
```json {filename="hugo.toml"}
[deployment]
[[deployment.targets]]
name: "(a name for your own reference)"

> URL: "s3://<BUCKET_NAME>?endpoint=https://<CELLAR_HOST>&region=fr-par"

Reviewer: ```suggestion
URL: "s3://<BUCKET_NAME>?endpoint=https://<CELLAR_HOST>&region=fr-par"
    ```

```

{{< /tab >}}

{{< /tabs >}}

- The deployment name is arbitrary and for your own reference: “production”, “test”, anything you’d like. 
- The cellar host address can found on your Clever Cloud console: select your Cellar and you’ll have the “Host” field. The address looks like: `cellar-c2.services.clever-cloud.com`

Save your configuration file: Hugo now knows where to send file when deploying.

To deploy, proceed as usual: `hugo` to build the website locally, then `hugo deploy` to push files through Cellar.
If all the steps have been correctly completed, you are now able to access your website with: 
 `https://<BUCKET_NAME>.<CELLAR_HOST>`

> <!--- // Gestion des certificats : si ca manque, voir le support pour forcer la gen d'un certificat. Sinon le faire en mode workaround dans la console. --->

Reviewer: ```suggestion

```

Is's possible to encounter an error where the website warns the user with a "not secure" message. 
This happens when the SSL certificate was not properly generated. 
You can manually generate one certificate by creating an application, then adding the domain name from the "domain" tab.
The certification propagation takes about 10-15mn. 

If you still encounter an issue at this point, please contact our support team. 

> <!--- // Ajouter une étape pour mettre en place le nom de domaine avec le bucket. Reprendre une partie de la page Cellar sur la section "Custom Domains" --->

Reviewer: ```suggestion

```

> <!--- //Rajouter une section prelim en début pour expliquer pourquoi passer par Cellar plutôt que runtime >

Reviewer: Why this?

> <!--- //Rajouter une section prelim en début pour expliquer pourquoi passer par Cellar plutôt que runtime >

Reviewer: Commented HTML seems to break the page interactivity and this is why your tabs weren't working.
```suggestion

```
